cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,381 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#include "UrdfParser.h"
+
+#include <XmlFile.h>
+#include <XmlNode.h>
+#include <XmlParser.h>
+
+#include "Assert/Assert.h"
+
+UrdfRobotDesc UrdfParser::parse(const std::string& file_name)
+{
+    FXmlFile file;
+    file.LoadFile(UTF8_TO_TCHAR(file_name.c_str()));
+
+    FXmlNode* robot_node = file.GetRootNode();
+    ASSERT(robot_node->GetTag().Equals(TEXT("robot")));
+
+    return parseRobotNode(robot_node);
+}
+
+UrdfRobotDesc UrdfParser::parseRobotNode(FXmlNode* robot_node)
+{
+    UrdfRobotDesc robot_desc;
+    // parse all nodes
+    for (auto& child_node : robot_node->GetChildrenNodes()) {
+        const FString& tag = child_node->GetTag();
+
+        if (tag.Equals(TEXT("link"))) {
+            UrdfLinkDesc link_desc = parseLinkNode(child_node);
+            ASSERT(!robot_desc.link_descs_.count(link_desc.name_));
+            robot_desc.link_descs_[link_desc.name_] = std::move(link_desc);
+        } else if (tag.Equals(TEXT("joint"))) {
+            UrdfJointDesc joint_desc = parseJointNode(child_node);
+            ASSERT(!robot_desc.joint_descs_.count(joint_desc.name_));
+            robot_desc.joint_descs_[joint_desc.name_] = std::move(joint_desc);
+        } else if (tag.Equals(TEXT("material"))) {
+            UrdfMaterialDesc material_desc = parseMaterialNode(child_node);
+            ASSERT(!robot_desc.material_descs_.count(material_desc.name_));
```

Comment:
```
            ASSERT(material_desc.name_ != "");
            ASSERT(!material_desc.is_reference_);
```

Reply:
Material node is discussed in previous comment.

Reply:
See my previous comments.